design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+
+We want to position ksqlDB as a JSON-friendly database. Support for reading, writing, and
+traversing data as JSON presents an important value-add. And users of traditional databases like
+PostgreSQL already enjoy powerful support for JSON.
+
+Today, ksqlDB can:
+* Read JSON-formatted keys and values in input records.
+* Serialize keys and values as JSON in output records.
+* Extract a string value at a given JSON path.
+
+
+## What is in scope
+
+* Validate if a given string contains JSON value
+* Calculate the length of a JSON array encoded in a string
+* Extract keys and records from a given JSON-formatted string
+* Concatenate JSON-formatted strings
+* Convert a ksqlDB record to JSON-formatted string
+
+## What is not in scope
+
+* A JSON datatype
+* Binary JSON support
+
+## Public APIs
+
+Users can use ksqlDB’s varchar type on columns that they know are JSON formatted. A JSON-formatted
+string can also be nested within a map or struct.
+
+### is_json
+
+```
+is_json(json_string)
+```
+
+Given a string, returns true if it can be parsed as a valid JSON value, false otherwise.
+
+#### Examples
+
+```
+is_json("[1, 2, 3]") // returns true
+is_json("{}") // returns true
+is_json("1") // returns true
+is_json("\"abc\"") // returns true
+is_json("null") // returns true
+is_json("") // returns false
+is_json("abc") // returns false
+is_json(NULL) // returns false
+```
+
+
+### json_array_length
+
+```
+json_array_length(json_string)
+```
+
+Given a string, parses it as a JSON value and returns the length of the top-level array. Returns
+`NULL` if the string can't be interpreted as a JSON array, i.e., it is `NULL` or it does not contain
+valid JSON, or the JSON value is not an array.
+
+#### Examples
+
+```
+json_array_length("[1, 2, 3]") // returns 3
+json_array_length("[1, [1, [2]], 3]") // returns 3
+json_array_length("[]") // returns 0
+json_array_length("{}") // returns NULL
+json_array_length("123") // returns NULL
+json_array_length("abc") // returns NULL
+json_array_length(NULL) // returns NULL
+```
+
+### json_keys
+
+```
+json_keys(json_string)
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of strings representing the
+top-level keys. Returns `NULL` if the string can't be interpreted as a JSON object, i.e., it is
+`NULL` or it does not contain valid JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_keys("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns ["a", "b", "c"]
+json_keys("{}") // returns []
+json_keys("[]") // returns NULL
+json_keys("") // returns NULL
+json_keys("123") // returns NULL
+json_keys("abc") // returns NULL
+json_keys(NULL) // returns NULL
+```
+
+### json_records
+
+```
+json_records(json_string)
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of two-element arrays
+representing the top-level keys and values. Returns `NULL` if the string can't be interpreted as a
```

Comment:
json objects don't have to have the same value type, right? what would happen if we passed in `{"a": 1, "b": "X"}` (subtext: ksqlDB arrays are strongly typed)

Reply:
Based on the example below, I think the result in ksql types would be: `Array[Array['a',  '1'], Array['b', '"X"']]`. Note: double-quotes surrounding the value of `b`. Essentially each inner value (meaning the second item in the nested arrays) in the result is equivalent to calling `to_json_string` on the JSON value input.

I assume the relevant the UDF's return type is `Array<Array<string>>`.

Reply:
Agree, the return type should be `Array<Array<string>>`.

Reply:
We should update the KLIP to call out the exact type of the result array.

Reply:
Btw: Why do we return an array of 2-element arrays? Would it be better to return an array of structs with two string field instead, ie, `Array<Struct<json_key:String,json_value:String>>`

Reply:
Good ideas @mjsax! I specified the return type of all functions in https://github.com/confluentinc/ksql/pull/8550/commits/145d932752706b0bf995d1ff7899d0d244dc14c1.

Reply:
Would be good to hear from @colinhicks what he thinks about the return type? Pros/Cons for both ideas? Would also be worth to add to rejected alternatives.

Reply:
Meanwhile I added a line on why an array of structs is a better alternative than an arrays of arrays - https://github.com/confluentinc/ksql/pull/8550/commits/391c9e04421a14c2dadfb0d4811ad11988b87773.

Reply:
@colinhicks @mjsax @agavra Apologies for a little bit of rehashing things.  While reviewing the PR for this, I asked if it'd make sense for the return type to be Map<String, String> instead of Array<Struct<Key,Key>>.  The types are isomorphic / contain the same information. 

My slight preference is for a Map, but that's based on how I think about JSON objects.  What do you all think?

Reply:
FWIW I don't have strong opinion on this. As Jim pointed out, the types are isomorphic, so I don't have any preference or arguments one way or another.

Reply:
I'm pro-Map. Thanks for the suggestion, @jnh5y!

Key-iteration order is also not guaranteed for JSON objects, so choosing Map seems better in that it would also eliminate any chance folks would expect to always see keys in the same order. Another benefit with map is that users can "pluck" a single value by key. Meanwhile, the results can still be iterated over with transform, filter, and reduce.

Reply:
I don't think `Map<String,String>` would work, as we could have duplicate keys that the `Map` cannot store?

Reply:
What about duplicate keys? A `Map` could not represent it.

Reply:
This SO answer helped me see that things could be tricky: https://stackoverflow.com/a/53529204

I don't know what to recommend here.  This is definitely a sharp edge.

Reply:
> What about duplicate keys?

Duplicat keys is either not an issue or a way bigger issue, depending on how we look at them. Jackson does not support duplicate keys, as `ObjectNode` maintains its fields in `LinkedHashMap` that does not allow duplicates ([link](https://github.com/FasterXML/jackson-databind/blob/279d2001db6435ea1e92c48d9e796baa74403dd7/src/main/java/com/fasterxml/jackson/databind/node/ObjectNode.java#L26)) and it is unlikely that they will support them in the near future (https://github.com/FasterXML/jackson-databind/issues/2809).

Reply:
It seems like the options are:
* Go with Array<Struct<String,String>> (since that's the existing code and it'd cover things if there was a desire to support the JSON spec to the letter).
* Switch to Map<String, String> (since this is what the library supports).

Thoughts?  At this point, I am unsure what to suggest.

Reply:
Might be a product call, if we should follow the spec or the limitations of Jackson... \cc @colinhicks

Reply:
We discussed this offline and agreed not to support duplicate keys for now and proceed with `Map<String, String>` return type here. I updated the doc in 85de9ab.